File: jscomp/test/bs_stack_test.js
'use strict';

var Mt = require("./mt.js");
var Caml_option = require("melange/lib/js/caml_option.js");
var Js_undefined = require("melange/lib/js/js_undefined.js");
var Belt_MutableQueue = require("melange/lib/js/belt_MutableQueue.js");
var Belt_MutableStack = require("melange/lib/js/belt_MutableStack.js");

var suites = {
  contents: /* [] */0
};

var test_id = {
  contents: 0
};

function eq(loc, x, y) {
  Mt.eq_suites(test_id, suites, loc, x, y);
}

function inOrder(v) {
  var current = v;
  var s = {
    root: undefined
  };
  var q = {
    length: 0,
    first: undefined,
    last: undefined
  };
  while(current !== undefined) {
    var v$1 = current;
    Belt_MutableStack.push(s, v$1);
    current = v$1.left;
  };
  while(s.root !== undefined) {
    current = Belt_MutableStack.popUndefined(s);
    var v$2 = current;
    Belt_MutableQueue.add(q, v$2.value);
    current = v$2.right;
    while(current !== undefined) {
      var v$3 = current;
      Belt_MutableStack.push(s, v$3);
      current = v$3.left;
    };
  };
  return Belt_MutableQueue.toArray(q);
}

function inOrder3(v) {
  var current = v;
  var s = {
    root: undefined
  };
  var q = {
    length: 0,
    first: undefined,
    last: undefined
  };
  while(current !== undefined) {
    var v$1 = current;
    Belt_MutableStack.push(s, v$1);
    current = v$1.left;
  };
  Belt_MutableStack.dynamicPopIter(s, (function (popped) {
          Belt_MutableQueue.add(q, popped.value);
          var current = popped.right;
          while(current !== undefined) {
            var v = current;
            Belt_MutableStack.push(s, v);
            current = v.left;
          };
        }));
  return Belt_MutableQueue.toArray(q);
}

function inOrder2(v) {
  var todo = true;
  var cursor = v;
  var s = {
    root: undefined
  };
  var q = {
    length: 0,
    first: undefined,
    last: undefined
  };
  while(todo) {
    if (cursor !== undefined) {
      var v$1 = cursor;
      Belt_MutableStack.push(s, v$1);
      cursor = v$1.left;
    } else if (s.root !== undefined) {
      cursor = Belt_MutableStack.popUndefined(s);
      var current = cursor;
      Belt_MutableQueue.add(q, current.value);
      cursor = current.right;
    } else {
      todo = false;
    }
  };
}

function n(l, r, a) {
  return {
          value: a,
          left: Js_undefined.fromOption(l),
          right: Js_undefined.fromOption(r)
        };
}

var test1 = n(Caml_option.some(n(Caml_option.some(n(undefined, undefined, 4)), Caml_option.some(n(undefined, undefined, 5)), 2)), Caml_option.some(n(undefined, undefined, 3)), 1);

function pushAllLeft(st1, s1) {
  var current = st1;
  while(current !== undefined) {
    var v = current;
    Belt_MutableStack.push(s1, v);
    current = v.left;
  };
}

var test2 = n(Caml_option.some(n(Caml_option.some(n(Caml_option.some(n(Caml_option.some(n(undefined, undefined, 4)), undefined, 2)), undefined, 5)), undefined, 1)), undefined, 3);

var test3 = n(Caml_option.some(n(Caml_option.some(n(Caml_option.some(n(undefined, undefined, 4)), undefined, 2)), undefined, 5)), Caml_option.some(n(undefined, undefined, 3)), 1);

eq("File \"bs_stack_test.ml\", line 137, characters 6-13", inOrder(test1), [
      4,
      2,
      5,
      1,
      3
    ]);

eq("File \"bs_stack_test.ml\", line 140, characters 6-13", inOrder3(test1), [
      4,
      2,
      5,
      1,
      3
    ]);

Mt.from_pair_suites("bs_stack_test.ml", suites.contents);

exports.suites = suites;
exports.test_id = test_id;
exports.eq = eq;
exports.inOrder = inOrder;
exports.inOrder3 = inOrder3;
exports.inOrder2 = inOrder2;
exports.n = n;
exports.test1 = test1;
exports.pushAllLeft = pushAllLeft;
exports.test2 = test2;
exports.test3 = test3;
/* test1 Not a pure module */

File: jscomp/test/bs_stack_test.ml
let suites :  Mt.pair_suites ref  = ref []
let test_id = ref 0
let eq loc x y = Mt.eq_suites ~test_id ~suites loc x y



type node = {
  value : int ;
  left : t;
  right : t
}
and t = node Js.undefined
[@@deriving jsProperties, getSet]



module S = Belt.MutableStack
module Q = Belt.MutableQueue

let inOrder (v : t) : int array =
  let current =  ref v in
  let s : node S.t = S.make () in
  let q : int Q.t = Q.make () in
  while !current != Js.undefined do
    let v = Js.Undefined.getUnsafe !current  in
    S.push s v;
    current := leftGet v;
  done ;
  while not (S.isEmpty s ) do
    current := S.popUndefined s ;
    let v = Js.Undefined.getUnsafe !current in
    Q.add q (valueGet v);
    current := rightGet v ;
    while !current != Js.undefined do
      let v = Js.Undefined.getUnsafe !current  in
      S.push s v;
      current := leftGet v;
    done ;
  done;
  Q.toArray q

let inOrder3 (v : t) : int array =
  let current =  ref v in
  let s : node S.t = S.make () in
  let q : int Q.t = Q.make () in
  while !current != Js.undefined do
    let v = Js.Undefined.getUnsafe !current  in
    S.push s v;
    current := leftGet v;
  done ;
  S.dynamicPopIter s begin fun  popped ->
    Q.add q (valueGet popped);
    let current = ref (rightGet popped) in
    while !current != Js.undefined do
      let v = Js.Undefined.getUnsafe !current in
      S.push s v;
      current := leftGet v
    done
  end;
  Q.toArray q

let inOrder2 (v : t) =
  let todo = ref true in
  let cursor = ref v in
  let s : node S.t = S.make () in
  let q : int Q.t = Q.make () in
  while !todo do
    if !cursor != Js.undefined then
      (
        let v = (Js.Undefined.getUnsafe !cursor) in
        S.push s v;
        cursor := leftGet v)
    else
      begin
        if not (S.isEmpty s) then
          (cursor := S.popUndefined s ;
           let current = Js.Undefined.getUnsafe !cursor in
           Q.add q (valueGet current);
           cursor := rightGet current)
        else
          todo := false
      end
  done

let n
    ?l ?r a =
  node   ~value:a
    ~left:(Js.Undefined.fromOption l)
    ~right:(Js.Undefined.fromOption r)


let test1 =
  n 1
    ~l:
      (n 2
         ~l:(n 4 )
         ~r:(n 5))
    ~r:(n 3)

let pushAllLeft st1 s1 =
  let current = ref st1 in
  while !current != Js.undefined do
    let v = Js.Undefined.getUnsafe !current  in
    S.push s1 v;
    current := leftGet v;
  done
;;


let test2 =
  n 3
  ~l:(
    n 1
    ~l:(
      n 5
      ~l:
      (n 2
      ~l:
      (n 4)
      )
    )
  )

let test3 =

    n 1
    ~l:(
      n 5
      ~l:
      (n 2
      ~l:
      (n 4)
      )
    )
  ~r:(n 3)

;; eq __LOC__
  (inOrder (Js.Undefined.return test1))
  [|4;2;5;1;3|]
;; eq __LOC__
  (inOrder3 (Js.Undefined.return test1))
  [|4;2;5;1;3|]

;; Mt.from_pair_suites __FILE__ !suites
